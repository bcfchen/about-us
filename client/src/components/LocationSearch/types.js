import PropTypes from "prop-types";
import Location from "../../models/Location";

export const propTypes = {
    currentLocation: PropTypes.instanceOf(Location),
    onLocationSelected: PropTypes.func.isRequired
};
