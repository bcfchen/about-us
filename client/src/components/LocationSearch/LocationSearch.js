import React from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import Select from "react-select";
import * as locationActions from "../../redux/actions/locationActions";
import * as weatherActions from "../../redux/actions/weatherActions";

class LocationSearch extends React.Component {
    constructor(props) {
        super(props);
    }

    onLocationSelected = selectedLocation => {
        if (selectedLocation) {
            let selectedLocationObj = this.props.locations.filter(loc => loc.key === selectedLocation.value)[0];
            this.props.weatherActions.loadFiveDayForecasts(selectedLocationObj.key);
            this.props.weatherActions.loadHourlyForecasts(selectedLocationObj);
        }
    }

    onInputChange = inputText => {
        if (inputText && inputText.length > 0) {
            this.props.locationActions.getSuggestedLocations(inputText);
        }
    }

    render() {
        const { hourlyForecasts, locations } = this.props;
        let currentLocation = hourlyForecasts[0].location;
        let currentVal = {
            value: currentLocation.key,
            label: currentLocation.getLocationString()
        };

        let locationOptions = locations.map(location => {
            return {
                value: location.key,
                label: location.getLocationString()
            }
        });

        return (
            <Select
                name="location"
                value={currentVal}
                onChange={this.onLocationSelected}
                onInputChange={this.onInputChange}
                options={locationOptions}
            />
        );
    }
}

const mapStateToProps = (state, ownProps) => {
    return {
        hourlyForecasts: state.weatherPage.hourlyForecasts,
        locations: state.weatherPage.locations
    };
};

const mapDispatchToProps = dispatch => {
    return {
        locationActions: bindActionCreators(locationActions, dispatch),
        weatherActions: bindActionCreators(weatherActions, dispatch)
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(LocationSearch);