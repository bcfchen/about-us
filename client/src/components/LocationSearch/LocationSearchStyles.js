export default theme => ({
    '@global': {
        '.Select-input': {
            display: 'inline-flex !important',
            padding: 0,
            height: 'auto',
            fontSize: "25px",
            color: "red",
            background: "transparent",
            marginRight: "auto",
            marginLeft: "auto",
            border: 0
        },
        ".Select-value-label": {
            marginRight: "auto",
            marginLeft: "auto",
            background: "transparent",
            border: 0,
            color: "red"
        },
        '.Select-input input': {
            background: 'transparent',
            border: 0,
            padding: 0,
            cursor: 'default',
            display: 'inline-block',
            fontFamily: 'inherit',
            fontSize: 'inherit',
            margin: 0,
            outline: 0,
            color: "white"
        },
        '.Select.is-focused:not(.is-open) > .Select-control': {
            boxShadow: 'none'
        },
        ".Select-arrow-zone": {
            display: "none"
        },
        ".Select-clear-zone": {
            display: "none"
        },
        ".Select-control": {
            background: "transparent",
            borderWidth: 0
        },
        ".Select.has-value.is-clearable.Select--single > .Select-control .Select-value ": {
            paddingLeft: 0,
            paddingRight: 0
        },
        "div.is-focused>.Select-control": {
            background: "transparent !important"
        }
    },
});