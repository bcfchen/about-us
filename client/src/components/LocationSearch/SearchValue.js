import React from "react";
import Typography from '@material-ui/core/Typography';

const SearchValue = ({ value, children }) => {
    return (
        <div className="Select-value async-select-value">
            <span className="Select-value-label">
                <Typography variant="headline" gutterBottom>{value.label}</Typography>
                {children}
            </span>
        </div>
    );
}

export default SearchValue;