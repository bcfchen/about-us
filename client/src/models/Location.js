class Location {
    constructor(city, region, country) {
        this.city = city;
        this.region = region;
        this.country = country;
    }

    getLocationString() {
        let hasInfo = this.city && this.region;
        return hasInfo ? this.city + ", " + this.region : "";
    }

    static fromLocationObj(locationObj) {
        return new Location(locationObj.city, locationObj.region, locationObj.country);
    }

    static fromLocationString(locationStr) {
        let locationArr = locationStr.split(", ");
        let city = locationArr[0],
            region = locationArr[1],
            country = locationArr[2];
        return new Location(city, region, country);
    }

    static fromLocationArr(locationArr) {
        let city = locationArr[1],
            region = locationArr[2],
            country = locationArr[3];
        return new Location(city, region, country);
    }
}

export default Location;