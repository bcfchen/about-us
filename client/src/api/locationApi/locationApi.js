import Location from "../../models/Location";
import { getJSONP } from "../baseApi/baseApi";
const config = require("../config.json");

export const getLocations = locationText => {
    let queryUrl = config.geoBytesAutoCompleteCities.basePath + locationText;
    return getJSONP(queryUrl)
        .then(locationsResponse => locationsResponse.map(location => Location.fromLocationString(location)));
}

export const getCurrentCityName = (position) => {
    return getCurrentPosition().then(coords => {
        let queryUrl = config.geoBytesNearbyCities.basePath + "longitude=" + coords.longitude + "&latitude=" + coords.latitude;
        return getJSONP(queryUrl)
    }).then(locationsResponse => Location.fromLocationArr(locationsResponse[0]).getCurrentLocationString());
}

const getCurrentPosition = () => {
    if (window.navigator.geolocation) {
        return new Promise((resolve, reject) => navigator.geolocation.getCurrentPosition(position => resolve(position.coords)))
    } else { return Promise.reject("no geo"); }
}