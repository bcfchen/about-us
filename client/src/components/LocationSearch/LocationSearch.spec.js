import React from "react";
import { shallow } from "enzyme";
import LocationSearch from "./LocationSearch";
import toast from "toastr";
import * as locationApi from "../../api/locationApi/locationApi";
import Location from "../../models/Location";
jest.mock("../../api/locationApi/locationApi");
jest.mock("toastr");

describe("<LocationSearch/>", () => {
    let component = null,
        defaultProps = {
            onLocationSelected: jest.fn(),
            currentLocation: new Location("San Francisco", "CA", "USA", "123")
        };
    let render = (props = defaultProps) => {
        component = shallow(<LocationSearch {...props} />).dive();
    };

    beforeEach(() => { render(); });
    afterEach(() => {
        jest.clearAllMocks();
    });

    describe("on page load", () => {
        it("should render correctly", () => {
            expect(component).toMatchSnapshot();
        });

        it("should assign correct location value", () => {
            expect(component.find("Async").prop("value").value).toEqual("123");
        });
    });

    describe("on location selected", () => {
        it("should trigger onLocationSelected", () => {
            component.find("Async").prop("onChange")();
            expect(defaultProps.onLocationSelected).toHaveBeenCalled();
        });
    });

    describe("on load location options", () => {
        describe("on load success", () => {
            it("should return correct locations for the selector", () => {
                let locationAvailable = new Location("San Francisco", "CA", "USA", "123");
                locationApi.getLocations.mockImplementation(() => Promise.resolve([locationAvailable]));
                component.find("Async").prop("loadOptions")().then(loadedOptions => {
                    expect(loadedOptions.options).toEqual([{ "label": "San Francisco, USA", "value": "123" }]);
                });
            });
        });

        describe("on load error", () => {
            it("should trigger toast error", () => {
                locationApi.getLocations.mockImplementation(() => Promise.reject("bad"));
                component.find("Async").prop("loadOptions")().then(loadedOptions => {
                    expect(toast.error).toHaveBeenCalled();
                });
            });
        });
    });
});