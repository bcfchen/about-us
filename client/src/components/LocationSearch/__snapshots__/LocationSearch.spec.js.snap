// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationSearch/> on page load should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LocationSearch
    classes={Object {}}
    currentLocation={
        Location {
            "city": "San Francisco",
            "country": "USA",
            "key": "123",
            "region": "CA",
          }
    }
    onLocationSelected={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "autoload": true,
      "cache": Object {},
      "children": [Function],
      "className": "async-select",
      "ignoreAccents": true,
      "ignoreCase": true,
      "loadOptions": [Function],
      "loadingPlaceholder": "Loading...",
      "name": "location",
      "onChange": [Function],
      "options": Array [],
      "searchPromptText": "Type to search",
      "value": Object {
        "label": "San Francisco, USA",
        "value": "123",
      },
      "valueComponent": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "autoload": true,
        "cache": Object {},
        "children": [Function],
        "className": "async-select",
        "ignoreAccents": true,
        "ignoreCase": true,
        "loadOptions": [Function],
        "loadingPlaceholder": "Loading...",
        "name": "location",
        "onChange": [Function],
        "options": Array [],
        "searchPromptText": "Type to search",
        "value": Object {
          "label": "San Francisco, USA",
          "value": "123",
        },
        "valueComponent": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
